examples/cevae.py

```diff
@@ -0,0 +1,65 @@
+import pyro
+from pyro import poutine
+from pyro.nn import PyroModule
+
+
+class Model(PyroModule):
+    num_treatments = 2
+
+    def forward(self, x, t=None, y=None, size=None):
+        if size is None:
+            size = x.size(0)
+        with pyro.plate("data", size, subsample=x):
+            z = pyro.sample("z", self.z_dist)
+            x = pyro.sample("x", self.x_dist(z), obs=x)
+            t = pyro.sample("t", self.t_dist(z), obs=t)
+            y = pyro.sample("y", self.y_dist(t, z), obs=y)
+        return y
+
+    def z_dist(self):
+        raise NotImplementedError
+
+    def x_dist(self, z):
+        raise NotImplementedError
+
+    def y_dist(self, t, y):
+        raise NotImplementedError
+
+
+class Guide(PyroModule):
+    num_treatments = 2
+
+    def forward(self, x, t=None, y=None, size=None):
+        if size is None:
+            size = x.size(0)
+        with pyro.plate("data", size, subsample=x):
+            t = pyro.sample("t", self.t_dist(x), obs=t)
+            y = pyro.sample("y", self.y_dist(t, x), obs=y)
+            pyro.sample("z", self.z_dist(t, x, y))
+
+    def t_dist(self, x):
+        raise NotImplementedError
+
+    def y_dist(self, t, y):
+        raise NotImplementedError
+
+    def z_dist(self, t, y, x):
+        raise NotImplementedError
+
+
+def ite(model, guide, x, num_samples=100):
```

Comment:
@eb8680 am I using `poutine.do` correctly here? I am literally transcribing
![image](https://user-images.githubusercontent.com/648532/68736049-72f24480-0594-11ea-9801-39ec596b0b86.png)

Reply:
If you want `ite` to correspond to the posterior predictive distributions in section 4 of the CEVAE paper then unfortunately no - `poutine.do` doesn't do backdoor adjustment, it just mutilates a model according to the definition of the do-operator.  

The expectations in the ITE are taken with respect to the *model*'s posterior marginal over `y | x, do(t=T)`, which is in turn computed by averaging over samples of the *marginal* distribution of `z` in the *guide*, which requires yet another marginalization over the auxiliary distributions over `y` and `t` in the *guide*.  

That is, to compute the ITE for the causal model in the paper, we do not intervene on `t` in the guide and we average over two distinct sets of samples from `y` - one from the guide and one from the model, which we return:
```py
guide_plate = pyro.plate('guide_particles', num_samples, dim=-2)
model_plate = pyro.plate('model_particles', num_samples, dim=-3)
with poutine.block(hide=['y', 't']), guide_plate:
    guide_trace = poutine.trace(poutine.block(guide, hide=['y', 't'])).get_trace(x)
with model_plate, guide_plate:
    y0 = poutine.replay(poutine.do(model, data=dict(t=0)), guide_trace)(x)
```

I believe these tedious and error-prone manipulations could be automated almost entirely if we implement a more complete semantics for counterfactuals, such as the formalism described in [Single World Intervention Graphs: A Primer](https://pdfs.semanticscholar.org/07bb/cb458109d2663acc0d098e8913892389a2a7.pdf).